internal/exp/metrics/staleness/priority_queue.go

```diff
@@ -0,0 +1,97 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package staleness
+
+import (
+	"container/heap"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+)
+
+type priorityQueueImpl []*queueItem
+
+type queueItem struct {
+	key   identity.Stream
+	prio  time.Time
+	index int
+}
+
+func (pq priorityQueueImpl) Len() int { return len(pq) }
+
+func (pq priorityQueueImpl) Less(i, j int) bool {
+	// We want Pop to give us the lowest priority
+	return pq[i].prio.Before(pq[j].prio)
+}
+
+func (pq priorityQueueImpl) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *priorityQueueImpl) Push(x any) {
+	n := len(*pq)
+	item := x.(*queueItem)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *priorityQueueImpl) Pop() any {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+func (pq *priorityQueueImpl) Update(item *queueItem, newPrio time.Time) {
+	item.prio = newPrio
+	heap.Fix(pq, item.index)
+}
+
+type PriorityQueue struct {
```

Comment:
wdyt about `PriorityQueue` (as used in the Staleness type) being an interface and having something like private like `heapQueue` be the actual implementation. Couples those two a little less

Reply:
Done